src/main/java/com/onarandombox/MultiverseCore/listeners/NewMVPlayerListener.java

```diff
@@ -0,0 +1,326 @@
+                    teleportee.getName(), event.getTo().getWorld());
+            return;
+        }
+        if (fromWorld.filter(world -> world.equals(toWorld)).isDefined()) {
+            // We don't want to do anything if the destination world is the same as the origin world.
+            Logging.fine("Player '%s' is teleporting to the same world.", teleportee.getName());
+            return;
+        }
+
+        ResultGroup worldEntryResult = worldEntryCheckerProvider.forWorld(teleportee, toWorld)
+                .canEnterWorld(fromWorld.getOrNull())
+                .onSuccess(() -> Logging.fine("MV-Core is allowing '%s' to go to '%s'.", teleportee.getName(), toWorld.getName()))
+                .onSuccessReason(EntryFeeResult.Success.ENOUGH_MONEY, () -> {
+                    economist.payEntryFee((Player) teleporter, toWorld);
+                    //TODO send payment message
+                })
+                .onFailure(() -> {
+                    event.setCancelled(true);
+                    Logging.fine("MV-Core is denying '%s' from going to '%s'.", teleportee.getName(), toWorld.getName());
+                });
+
+        Logging.fine("World entry result for player '%s', from '%s' to '%s': %s",
+                teleportee.getName(), fromWorld.map(MVWorld::getName).getOrNull(), toWorld.getName(), worldEntryResult);
+
+        sendWorldEntryMessage(teleporter, teleportee, fromWorld.getOrNull(), toWorld, worldEntryResult);
+    }
+
+    private void sendWorldEntryMessage(CommandSender teleporter, Player teleportee, MVWorld fromWorld, MVWorld toWorld, ResultGroup worldEntryResult) {
+        if (worldEntryResult.isSuccess()) {
+            return;
+        }
+
+        if (teleportee.equals(teleporter)) {
+            teleporter.sendMessage("You are unable to teleport to world '%s'".formatted(toWorld.getName()));
+        } else {
+            teleporter.sendMessage("You are unable to teleport '%s' to world '%s'".formatted(teleportee.getName(), toWorld.getName()));
+        }
+
+        worldEntryResult
+                .onFailureReason(WorldAccessResult.Failure.class, reason -> {
+                    teleporter.sendMessage("You do not have permission to access world '%s'".formatted(toWorld.getName()));
+                })
+                .onFailureReason(PlayerLimitResult.Failure.class, reason -> {
+                    teleporter.sendMessage("World '%s' is full".formatted(toWorld.getName()));
+                })
+                .onFailureReason(EntryFeeResult.Failure.class, reason -> {
+                    teleporter.sendMessage("You do not have enough money to pay world '%s' entry fee.".formatted(toWorld.getName()));
+                    //TODO Formatted entry fee amount
+                })
+                .onFailureReason(BlacklistResult.Failure.class, reason -> {
+                    teleporter.sendMessage("World '%s' is blacklisted by world '%s'".formatted(fromWorld.getName(), toWorld.getName()));
+                });
```

Comment:
This is what has been throwing me off. Like, what if a new failure reason is added and used elsewhere and forgotten here? Is there some way we can avoid that kind of situation?